xdsl/irdl/dominance.py

```diff
@@ -0,0 +1,91 @@
+from xdsl.ir import Block, Region
+
+
+class DominanceInfo:
+    """
+    Computes and exposes the dominance relation amongst blocks of a region.
+
+    https://en.wikipedia.org/w/index.php?title=Dominator_(graph_theory)&oldid=1189814332
+    """
+
+    _dominfo: dict[Block, set[Block]]
+
+    def __init__(self, region: Region):
+        """
+        Compute (improper) dominance.
+
+        https://en.wikipedia.org/w/index.php?title=Dominator_(graph_theory)&oldid=1189814332
+        """
+
+        self._dominfo = {}
+
+        # No block, no work
+        if not (region.blocks):
+            return
+
+        # Build the preceding relationship
+        pred: dict[Block, set[Block]] = {}
+        for b in region.blocks:
+            pred[b] = set()
+        for b in region.blocks:
+            if b.last_op is not None:
+                for s in b.last_op.successors:
+                    pred[s].add(b)
+
+        # Get entry and other blocks
+        entry, *blocks = region.blocks
+
+        # The entry block is only dominated by itself
+        self._dominfo[entry] = {entry}
+
+        # Instantiate other blocks dominators to all blocks
+        for b in blocks:
+            self._dominfo[b] = set(region.blocks)
+
+        # Iteratively filter out dominators until it converges
+        changed = True
+        while changed:
+            changed = False
+            for b in blocks:
+                oldie = self._dominfo[b].copy()
+                self._dominfo[b] = {b} | (
+                    set[Block].intersection(*(self._dominfo[p] for p in pred[b]))
+                    if pred[b]
+                    else set()
+                )
+                if oldie != self._dominfo[b]:
+                    changed = True
+
+    def properly_dominates(self, a: Block, b: Block) -> bool:
```

Comment:
```suggestion
    def strictly_dominates(self, a: Block, b: Block) -> bool:
```
I think strict is the technical term here, for the transitive closure of a relation that doesn't include the source

Reply:
Yeah but MLIR uses properly; do you prefer strictly?

Reply:
It's just a bit more natural for me, it's like strict relations (https://en.wikipedia.org/wiki/Partially_ordered_set), it immediately implies non-reflexivity. Maybe there's a term of art for dominance that would overrule this if all the dominance experts will get confused by an unfamiliar term :)

Reply:
Sure, will use strict, this is just picking who gets confused though :wink: 
I like choosing the mathematical consensus too, though :+1: